### Note
* Instructions have been included for each segment. You do not have to follow them exactly, but they are included to help you think through the steps.

In [192]:
# Dependencies and Setup
import pandas as pd
import numpy as np
import csv

# File to Load (Remember to Change These)
file_to_load = "Resources/purchase_data.csv"

# Read Purchasing File and store into Pandas data frame
purchase_data = pd.read_csv(file_to_load)
# Preview of DataFrame
purchase_data.head(10)

Purchase ID             SN  ...                                  Item Name Price
0            0        Lisim78  ...  Extraction, Quickblade Of Trembling Hands  3.53
1            1    Lisovynya38  ...                          Frenzied Scimitar  1.56
2            2     Ithergue48  ...                               Final Critic  4.88
3            3  Chamassasya86  ...                                Blindscythe  3.27
4            4      Iskosia90  ...                                       Fury  1.44
5            5        Yalae81  ...                                  Dreamkiss  3.61
6            6      Itheria73  ...     Interrogator, Blood Blade of the Queen  2.18
7            7    Iskjaskst81  ...                              Abyssal Shard  2.67
8            8      Undjask33  ...                                  Souleater  1.10
9            9    Chanosian48  ...               Ghastly Adamantite Protector  3.58

[10 rows x 7 columns]

In [193]:
# Check for missing data 
purchase_data.count()

Purchase ID    780
SN             780
Age            780
Gender         780
Item ID        780
Item Name      780
Price          780
dtype: int64

* No Missing Values, Dataset looks good

## Player Count

* Display the total number of players


In [194]:
# Total Number of Players 
total_players = purchase_data['SN'].unique()
players_count = len(total_players)
print(f'The total number of players is : {players_count}')


The total number of players is : 576


## Purchasing Analysis (Total)

* Run basic calculations to obtain number of unique items, average price, etc.


* Create a summary data frame to hold the results


* Optional: give the displayed data cleaner formatting


* Display the summary data frame


In [195]:
# Count of unique items
unique_items = len(purchase_data['Item ID'].unique())

In [196]:
# the average price
avg_price= purchase_data['Price'].mean()

In [197]:
# Count the number of purchases
total_purchases= len(purchase_data['Purchase ID'])

In [198]:
# Calculate the total revenue from all sales
total_revenue= sum(purchase_data['Price'])

In [199]:
# create dataframe to hold above results
purchase_summary_df = pd.DataFrame({
    'Number of Unique items': unique_items,
    'Average Price': '${:.2f}'.format(avg_price),
    'Number of Purchases': total_purchases,
    'Total Revenue': '${:.2f}'.format(total_revenue)
},index=[0])
purchase_summary_df

Number of Unique items Average Price  Number of Purchases Total Revenue
0                     179         $3.05                  780      $2379.77

## Gender Demographics

* Percentage and Count of Male Players


* Percentage and Count of Female Players


* Percentage and Count of Other / Non-Disclosed




In [200]:
# Gender Count
purchase_data['Gender'].value_counts()


Male                     652
Female                   113
Other / Non-Disclosed     15
Name: Gender, dtype: int64

In [201]:
# Check for Duplicate nnames
gender_groups = purchase_data.groupby(['Gender'])
count = gender_groups["SN"].value_counts()
count


Gender                 SN         
Female                 Chamjask73     3
                       Ialallo29      3
                       Phyali88       3
                       Tyidaim51      3
                       Umolrian85     3
                                     ..
Other / Non-Disclosed  Haerithp41     1
                       Jiskirran77    1
                       Lirtim36       1
                       Rairith81      1
                       Sundim98       1
Name: SN, Length: 576, dtype: int64

In [202]:
# Remove duplicate names
unique_gender_df = purchase_data.drop_duplicates('SN')
# Gender count for each unique name
gender_count = unique_gender_df['Gender'].value_counts()
gender_count

Male                     484
Female                    81
Other / Non-Disclosed     11
Name: Gender, dtype: int64

In [203]:
# Get percentage of player gender
gender_percent = (unique_gender_df['Gender'].value_counts()/players_count)*100
gender_percent

Male                     84.027778
Female                   14.062500
Other / Non-Disclosed     1.909722
Name: Gender, dtype: float64

In [204]:
# Write into a DataFrame 
gender_demo_df = pd.DataFrame({
    'Total Count': gender_count,
    'Percentage of Players': gender_percent.map('{:.2f}%'.format)
})
gender_demo_df

Total Count Percentage of Players
Male                           484                84.03%
Female                          81                14.06%
Other / Non-Disclosed           11                 1.91%


## Purchasing Analysis (Gender)

* Run basic calculations to obtain purchase count, avg. purchase price, avg. purchase total per person etc. by gender




* Create a summary data frame to hold the results


* Optional: give the displayed data cleaner formatting


* Display the summary data frame

In [205]:
# Use groupby to group data by Gender 
unique_gender_count = gender_groups['SN'].nunique()
gender_groups.describe()

Purchase ID                          ... Price             
                            count        mean         std  ...   50%    75%   max
Gender                                                     ...                   
Female                      113.0  379.380531  211.605484  ...  3.45  4.230  4.90
Male                        652.0  392.516871  227.516414  ...  3.09  4.080  4.99
Other / Non-Disclosed        15.0  334.600000  234.524991  ...  3.45  3.875  4.75

[3 rows x 32 columns]

In [206]:
# purchase Count
purchase_count = gender_groups['Gender'].count()
purchase_count

Gender
Female                   113
Male                     652
Other / Non-Disclosed     15
Name: Gender, dtype: int64

In [207]:
# Average Purchase price
purchase_mean = gender_groups['Price'].mean()
purchase_mean

Gender
Female                   3.203009
Male                     3.017853
Other / Non-Disclosed    3.346000
Name: Price, dtype: float64

In [208]:
# total revenue
purchase_total = gender_groups['Price'].sum()
purchase_total 

Gender
Female                    361.94
Male                     1967.64
Other / Non-Disclosed      50.19
Name: Price, dtype: float64

In [209]:
# Average purchase per person
avg_purchase_per_person = purchase_total/unique_gender_count
avg_purchase_per_person

Gender
Female                   4.468395
Male                     4.065372
Other / Non-Disclosed    4.562727
dtype: float64

In [210]:
# Write into a summary dataframe
summary_df = pd.DataFrame({"Purchase Count": purchase_count, "Average Purchase Price": purchase_mean, 
                           "Total Purchase Value": purchase_total, "Average Total Purchase per Person": avg_purchase_per_person})
summary_df["Average Purchase Price"] = summary_df["Average Purchase Price"].map("${:.2f}".format)
summary_df["Average Total Purchase per Person"] = summary_df["Average Total Purchase per Person"].map("${:.2f}".format)
summary_df["Total Purchase Value"] = summary_df["Total Purchase Value"].map("${:,.2f}".format)

purchase_analysis_gender_df = summary_df[["Purchase Count", "Average Purchase Price", "Total Purchase Value", 
                         "Average Total Purchase per Person"]]
purchase_analysis_gender_df

Purchase Count  ... Average Total Purchase per Person
Gender                                 ...                                  
Female                            113  ...                             $4.47
Male                              652  ...                             $4.07
Other / Non-Disclosed              15  ...                             $4.56

[3 rows x 4 columns]

## Age Demographics

* Establish bins for ages


* Categorize the existing players using the age bins. Hint: use pd.cut()


* Calculate the numbers and percentages by age group


* Create a summary data frame to hold the results


* Optional: round the percentage column to two decimal points


* Display Age Demographics Table


In [211]:
# Create bins
bins = [0, 9, 14, 19, 24, 29, 34, 39, 100]
demographics = ["<10", "10-14", "15-19", "20-24", "25-29", "30-34", "35-39", "40+"]

In [212]:
# Cut the data into categories
purchase_data["Age Demographic"] = pd.cut(purchase_data["Age"], bins, labels = demographics)
age_demographics = purchase_data.groupby("Age Demographic")


In [213]:
# CAlculate the percentage of players and number of plers by demographics
a_total_count = age_demographics["SN"].nunique()
a_total_players = purchase_data["SN"].nunique()
a_percentage_of_players = (a_total_count/a_total_players) * 100

In [214]:
# Write a summary data frame of age demographic 
a_summary_df = pd.DataFrame({"Total Count": a_total_count, "Percentage of Players": a_percentage_of_players})
a_summary_df["Percentage of Players"] = a_summary_df["Percentage of Players"].map("{:.2f}%".format)
age_demo_df = a_summary_df[["Total Count", "Percentage of Players"]]
age_demo_df

Total Count Percentage of Players
Age Demographic                                   
<10                       17                 2.95%
10-14                     22                 3.82%
15-19                    107                18.58%
20-24                    258                44.79%
25-29                     77                13.37%
30-34                     52                 9.03%
35-39                     31                 5.38%
40+                       12                 2.08%

## Purchasing Analysis (Age)

* Bin the purchase_data data frame by age


* Run basic calculations to obtain purchase count, avg. purchase price, avg. purchase total per person etc. in the table below


* Create a summary data frame to hold the results


* Optional: give the displayed data cleaner formatting


* Display the summary data frame

In [215]:

# Purchase count per demographic
pa_purchase_count = age_demographics["Purchase ID"].count() 

# Average price per demographic 
pa_avg_price = age_demographics["Price"].mean() 

 # Total Amount purchased per demographic 
pa_total_purchase_value = pa_purchase_count * pa_avg_price
# Average purchase total per person
avg_total_per_person = age_demographics["Price"].sum()/age_demographics["SN"].nunique()

pa_summary_df = pd.DataFrame({"Purchase Count": pa_purchase_count, "Average Purchase Price": pa_avg_price,
                          "Total Purchase Value": pa_total_purchase_value, "Average Total per Person": avg_total_per_person})

pa_summary_df["Average Purchase Price"] = pa_summary_df["Average Purchase Price"].map("${:.2f}".format)
pa_summary_df["Average Total per Person"] = pa_summary_df["Average Total per Person"].map("${:.2f}".format)
pa_summary_df["Total Purchase Value"] = pa_summary_df["Total Purchase Value"].map("${:.2f}".format)

purchase_analysis_age_df = pa_summary_df[["Purchase Count", "Average Purchase Price", "Total Purchase Value", 
                         "Average Total per Person"]]
purchase_analysis_age_df

Purchase Count  ... Average Total per Person
Age Demographic                  ...                         
<10                          23  ...                    $4.54
10-14                        28  ...                    $3.76
15-19                       136  ...                    $3.86
20-24                       365  ...                    $4.32
25-29                       101  ...                    $3.81
30-34                        73  ...                    $4.12
35-39                        41  ...                    $4.76
40+                          13  ...                    $3.19

[8 rows x 4 columns]

## Top Spenders

* Run basic calculations to obtain the results in the table below


* Create a summary data frame to hold the results


* Sort the total purchase value column in descending order


* Optional: give the displayed data cleaner formatting


* Display a preview of the summary data frame



In [216]:
# Count purchases made by unique SN
spender_count = purchase_data["SN"].value_counts()

#Calculate the average purchase by unique SN
spender_average_purchase = purchase_data["Price"].groupby(purchase_data["SN"]).mean()

# Calculate the total purchase from each unique SN
spender_total_purchase = purchase_data["Price"].groupby(purchase_data["SN"]).sum()

# create dataframe
top_spenders = pd.DataFrame({
    "Purchase Count":spender_count,
    "Average Purchase Price":spender_average_purchase,
    "Total Purchase Value":spender_total_purchase
})

#  sort and format
top_spenders_sorted = top_spenders.sort_values("Total Purchase Value", ascending= False)
top_spenders_sorted["Average Purchase Price"] = top_spenders_sorted["Average Purchase Price"].map("${:.2f}".format)
top_spenders_sorted["Total Purchase Value"] = top_spenders_sorted["Total Purchase Value"].map("${:.2f}".format)

# show top 5
top_spenders_sorted.head(5)

Purchase Count Average Purchase Price Total Purchase Value
Lisosia93                 5                  $3.79               $18.96
Idastidru52               4                  $3.86               $15.45
Chamjask73                3                  $4.61               $13.83
Iral74                    4                  $3.40               $13.62
Iskadarya95               3                  $4.37               $13.10

## Most Popular Items

* Retrieve the Item ID, Item Name, and Item Price columns


* Group by Item ID and Item Name. Perform calculations to obtain purchase count, item price, and total purchase value


* Create a summary data frame to hold the results


* Sort the purchase count column in descending order


* Optional: give the displayed data cleaner formatting


* Display a preview of the summary data frame



In [217]:
# retrieve columns in new dataframe
popular_items_df = purchase_data[['Item ID', 'Item Name', 'Price']]

# count transactions for each item
items_purchase_count = popular_items_df.groupby(["Item ID","Item Name"])["Price"].count()

# Price of each item
item_price = popular_items_df.groupby(["Item ID","Item Name"])["Price"].mean()

# Calculate total revenue from each item
total_purchase_value = popular_items_df.groupby(["Item ID","Item Name"])["Price"].sum()

# Create summary data frame
pi_summary_df = pd.DataFrame({"Purchase Count": items_purchase_count, "Item Price": item_price, 
                           "Total Purchase Value": total_purchase_value})

# sort dataframe
pi_summary_df_sorted = pi_summary_df.sort_values(by="Purchase Count", ascending = False)

#format dataframe
pi_summary_df_sorted["Item Price"] = pi_summary_df_sorted["Item Price"].map("${:.2f}".format)
pi_summary_df_sorted["Total Purchase Value"] = pi_summary_df_sorted["Total Purchase Value"].map("${:.2f}".format)

# show summary data frame
pi_summary_df_sorted = pi_summary_df_sorted[["Purchase Count", "Item Price", "Total Purchase Value"]]
pi_summary_df_sorted.head()

Purchase Count  ... Total Purchase Value
Item ID Item Name                                                     ...                     
92      Final Critic                                              13  ...               $59.99
178     Oathbreaker, Last Hope of the Breaking Storm              12  ...               $50.76
145     Fiery Glass Crusader                                       9  ...               $41.22
132     Persuasion                                                 9  ...               $28.99
108     Extraction, Quickblade Of Trembling Hands                  9  ...               $31.77

[5 rows x 3 columns]

## Most Profitable Items

* Sort the above table by total purchase value in descending order


* Optional: give the displayed data cleaner formatting


* Display a preview of the data frame



In [218]:
# sort above table by total purchase
popular_items_sorted_purchase = pi_summary_df.sort_values("Total Purchase Value", ascending= False)

# format the table
popular_items_sorted_purchase["Item Price"]= popular_items_sorted_purchase["Item Price"].map("${:.2f}".format)
popular_items_sorted_purchase["Total Purchase Value"]= popular_items_sorted_purchase["Total Purchase Value"].map("${:.2f}".format)

#print the preview of the data frame
popular_items_sorted_purchase.head()


Purchase Count  ... Total Purchase Value
Item ID Item Name                                                     ...                     
92      Final Critic                                              13  ...               $59.99
178     Oathbreaker, Last Hope of the Breaking Storm              12  ...               $50.76
82      Nirvana                                                    9  ...               $44.10
145     Fiery Glass Crusader                                       9  ...               $41.22
103     Singed Scalpel                                             8  ...               $34.80

[5 rows x 3 columns]

## Three Observable Trends in Heroes of Pymoli

* Majority of our Players are males
* The 20-24 year old age group contributes approximately 45% of player count and spend more than any of the other age groups.

* "Final Critic" is both our most profitable and most popular item